### ISSUE TITLE
Wrong Best Value Report

### ISSUE THREAD

**Issue by etetteh** [optuna/optuna#2733]
I am using the optuna v2.8.0, but the following results does not look like what I expect.
The optimization is to maximize the average ROCAUC, and I don't understand why the value 11.2961 from epoch 1 is the best.

![Screenshot from 2021-06-07 18-18-48](https://user-images.githubusercontent.com/28512232/121094554-d36f5500-c7bc-11eb-9fd8-6e98e35a5768.png)

**Comment by hvy** [comment]
Not sure I'm interpreting your log correctly, but I understand that the value 11.2961 is one of your validation losses. Could you possibly share some of the code from your objective function. Do you happen to use pruning (or `Trial.report`)? You might get some unintended behavior if you're using one metric for pruning, that is validation loss, and another metric, that is average ROCAUC for the actual optimization.

Also, I'm curious if this is something that started to happen after upgrading to v2.8. Was this never an issue with earlier versions?

Thanks.

**Comment by etetteh** [comment]
I will post the entire code so you have a look.
```
def baseline_train(trial, model, weight_decay, optimizer, scheduler, num_epochs):            
    path = f"{split}_checkpoint/{trial.number}"
    os.makedirs(path, exist_ok=True)
    
    # Setting the seed
    np.random.seed(cfg.seed)
    random.seed(cfg.seed)
    torch.manual_seed(cfg.seed)

    if cfg.cuda:
        torch.cuda.manual_seed_all(cfg.seed)
        torch.backends.cudnn.deterministic = True
        torch.backends.cudnn.benchmark = False

    # Checkpointing
    start_epoch = 0
    best_metric = 0.
    weights_for_best_validauc = copy.deepcopy(model.state_dict())
    opt_weights_for_best_validauc = copy.deepcopy(optimizer.state_dict())
    metrics = []    
    if "checkpoint_path" in trial.user_attrs:
        checkpoint = torch.load(trial.user_attrs["checkpoint_path"])
        start_epoch = checkpoint["epoch"] + 1
        model.load_state_dict(checkpoint["model_state_dict"])
        optimizer.load_state_dict(checkpoint["optimizer_state_dict"])
        best_metric = checkpoint["best_metric"]
        
        weights_for_best_validauc = copy.deepcopy(model.state_dict())
        opt_weights_for_best_validauc = copy.deepcopy(optimizer.state_dict())

    for epoch in range(start_epoch, num_epochs):
        np.random.seed(cfg.seed + epoch)
        train_loss = train_epoch(
                               num_batches=cfg.num_batches,
                               epoch=epoch, 
                               model=model, 
                               device=device, 
                               train_loaders=train_loaders,
                               weight_decay=weight_decay,
                               optimizer=optimizer,
                               scheduler=scheduler,
                               criterion=train_criterion,)
        
        valid_loss, valid_auc = valid_epoch(
                                         name='Valid',
                                         epoch=epoch,
                                         model=model,
                                         device=device,
                                         data_loader=valid_loader,
                                         criterion=criterion,
                                         limit=cfg.num_batches//2)

        if np.mean(valid_auc) > best_metric:
            best_metric = np.mean(valid_auc)
            weights_for_best_validauc = copy.deepcopy(model.state_dict())
            optimizer_for_best_validauc = optimizer.state_dict()
        
 
        trial.report(valid_auc, epoch)
        
        torch.save(
            {
                "epoch": epoch,
                "model_state_dict": weights_for_best_validauc,
                "optimizer_state_dict": optimizer_for_best_validauc,
                "best_metric": best_metric,
            },
            os.path.join(path, split+"_model.pt"),
        )
        
        if trial.should_prune():
            raise optuna.TrialPruned()
    
    model.load_state_dict(weights_for_best_validauc)
    optimizer.load_state_dict(opt_weights_for_best_validauc)

    return best_metric   

criterion = torch.nn.BCEWithLogitsLoss()

def restart_from_checkpoint(study, trial):
    # Enqueue trial with the same parameters as the stale trial to use saved information.

    path = f"{split}+_checkpoint/{trial.number}/{split}+_model.pt"
    user_attrs = copy.deepcopy(trial.user_attrs)
    if os.path.exists(path):
        user_attrs["checkpoint_path"] = path

    study.add_trial(
        optuna.create_trial(
            state=optuna.trial.TrialState.WAITING,
            params=trial.params,
            distributions=trial.distributions,
            user_attrs=user_attrs,
            system_attrs=trial.system_attrs,
        )
    )


def objective(trial):
    params = {
        "lr": trial.suggest_categorical('lr', [1e-4, 1e-3, 1e-2]),
        "wd": trial.suggest_categorical('wd', [1e-6, 1e-5, 1e-4, 1e-3]),
        "dp_out": trial.suggest_float('dp_out', 0.0, 0.3, step=0.1),
        "amsgrad": trial.suggest_categorical('amsgrad', [True, False]), 
    }
    
    optimizer = torch.optim.Adam(model.parameters(), lr=params["lr"], amsgrad=params["amsgrad"]) 
    weight_decay = params["wd"]
    scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=30, gamma=0.1)
    best_auc = baseline_train(trial, model, weight_decay, optimizer, scheduler, num_epochs=cfg.num_epochs)
        
    return best_auc

if __name__ == "__main__":
    storage = optuna.storages.RDBStorage(
        "sqlite:///" + split + ".db", heartbeat_interval=1, failed_trial_callback=restart_from_checkpoint
    )
    
    sampler = optuna.samplers.TPESampler(seed=cfg.seed)
    study = optuna.create_study(
        sampler=sampler,
        pruner=optuna.pruners.MedianPruner(
            n_startup_trials=3, n_warmup_steps=5, interval_steps=3
            ),
        storage=storage, study_name=split + "_checkpoint", direction="maximize", load_if_exists=True
        )

    max_num_trials = cfg.num_trials
    num_rest_trials = max_num_trials - len(study.get_trials())
    study.optimize(func=objective, n_trials=num_rest_trials, gc_after_trial=True)

    print(f"Best value: {study.best_value}")
    print(f"Best params:\n {study.best_params}")

    df = study.trials_dataframe(attrs=("number", "value", "params", "state"))
    df.to_csv(split+"_trials.csv")

    with open(split+"_best_params.csv", 'w') as csvfile:
        field_names = ['Best Valid AUC', 'Best params']
        writer = csv.DictWriter(csvfile, fieldnames=field_names)

        writer.writeheader()
        writer.writerow({"Best Valid AUC": study.best_value, 
                         "Best params": study.best_params, 
                        })
```

**Comment by hvy** [comment]
At a quick glance, couldn't see anything wrong. Besides maybe taking the mean of `valid_auc`. If `valid_auc` is an array with more than one element, `trial.report` should raise an error. Could you do some print debugging of e.g. the values of `valid_loss`, `valid_auc` upon returning from `valid_epoch`, and check when you are pruning and/or returning from the objective function with `best_metric`?

**Comment by etetteh** [comment]
Sorry, `valid_auc` is not an array. I will remove the mean.
The goal is to search hyperparameters that `maximizes` the `valid_auc`. However, from the outputs, I am getting, there are trials and even epochs that have higher `valid_auc`, but the best output report is of trials with the worst or largest loss

**Comment by hvy** [comment]
Just to double check, it's not that you're mixing up `valid_auc` and `valid_loss`? It seems like the losses are told to optuna and optuna tries to maximize this loss.

**Comment by etetteh** [comment]
That was my guess too, and in fact, that is what is happening. Is there a fix to this bug

**Comment by etetteh** [comment]
My bad. I changed the `maximize` to `minimize`, but upon seeing your comment, I figured I have interchanged them. Thanks, and sorry

**Comment by etetteh** [comment]
We close this issue.

**Comment by hvy** [comment]
Good to hear!
